File: CoreFlows/examples/Python/SinglePhase/SinglePhase_1DDepressurisation.py
#!/usr/bin/env python
# -*-coding:utf-8 -*

import CoreFlows as cf
import cdmath as cm
import math 

def SinglePhase_1DDepressurisation():

	spaceDim = 1;
    # Prepare for the mesh
	print("Building mesh " );
	xinf = 0 ;
	xsup=4.2;
	nx=50;
	M=cm.Mesh(xinf,xsup,nx)

    # set the initial field
	initialPressure=155e5;
	initialVelocityX=0;
	initialTemperature=573;

   # set the boundary data for each boundary
	outletPressure=80e5;
	wallVelocityX=0;
	wallTemperature=573;

 	myProblem = cf.SinglePhase(cf.Liquid,cf.around155bars600K,spaceDim);
	nVar =  myProblem.getNumberOfVariables();

    # Prepare for the initial condition
	VV_Constant =[0]*nVar;

	# constant vector
	VV_Constant[0] = initialPressure ;
	VV_Constant[1] = initialVelocityX;
	VV_Constant[2] = initialTemperature ;


    #Initial field creation
	print("Building initial data" ); 
	myProblem.setInitialFieldConstant( spaceDim, VV_Constant, xinf, xsup, nx,"wall","outlet");

    # set the boundary conditions
	myProblem.setWallBoundaryCondition("wall", wallTemperature, wallVelocityX);
	myProblem.setOutletBoundaryCondition("outlet", outletPressure,[xsup]);


    # set the numerical method
	myProblem.setNumericalScheme(cf.upwind, cf.Explicit);
    
    # name of result file
	fileName = "1DDepressurisation";

    # simulation parameters 
	MaxNbOfTimeStep = 3 ;
	freqSave = 1;
	cfl = 0.95;
	maxTime = 500;
	precision = 1e-6;

	myProblem.setCFL(cfl);
	myProblem.setPrecision(precision);
	myProblem.setMaxNbOfTimeStep(MaxNbOfTimeStep);
	myProblem.setTimeMax(maxTime);
	myProblem.setFreqSave(freqSave);
	myProblem.setFileName(fileName);
	myProblem.setNewtonSolver(precision,20);
	#myProblem.saveConservativeField(True);
	if(spaceDim>1):
		myProblem.saveVelocity();
		pass
 
    # evolution
	myProblem.initialize();

	ok = myProblem.run();
	if (ok):
		print( "Simulation python " + fileName + " is successful !" );
		pass
	else:
		print( "Simulation python " + fileName + "  failed ! " );
		pass

	print( "------------ End of calculation !!! -----------" );

	myProblem.terminate();
	return ok

if __name__ == """__main__""":
    SinglePhase_1DDepressurisation()

File: CoreFlows/examples/C/SinglePhase_1DDepressurisation.cxx
#include "SinglePhase.hxx"

using namespace std;

int main(int argc, char** argv)
{
	//Preprocessing: mesh and group creation
	cout << "Building cartesian mesh" << endl;
	double xinf=0.0;
	double xsup=4.2;
	int nx=50;
	Mesh M(xinf,xsup,nx);
	double eps=1.E-8;
	M.setGroupAtPlan(xsup,0,eps,"Outlet");
	M.setGroupAtPlan(xinf,0,eps,"Wall");
	int spaceDim = M.getSpaceDimension();

	// set the initial field
	double initialPressure=155e5;
	double initialVelocityX=0;
	double initialTemperature=573;

	//set the boundary data for each boundary
	double outletPressure=80e5;
	double wallVelocityX=0;
	double wallTemperature=573;

	SinglePhase  myProblem(Liquid,around155bars600K,spaceDim);
	int nbPhase = myProblem.getNumberOfPhases();

	// Prepare for the initial condition
	int nVar = myProblem.getNumberOfVariables();
	Vector VV_constant(nVar);
	VV_constant(0) = initialPressure ;
	VV_constant(1) = initialVelocityX;
	VV_constant(2) = initialTemperature	;

	cout << "Building initial data" << endl;
	Field VV("Primitive", CELLS, M, nVar);

	myProblem.setInitialFieldConstant(M,VV_constant);

	//set the boundary conditions
	myProblem.setWallBoundaryCondition("Wall", wallTemperature, wallVelocityX);
	myProblem.setOutletBoundaryCondition("Outlet", outletPressure,vector<double>(1,xsup));

	// set the numerical method
	myProblem.setNumericalScheme(upwind, Explicit);
	myProblem.setEntropicCorrection(true);

	// name file save
	string fileName = "1DDepressurisation";

	// parameters calculation
	unsigned MaxNbOfTimeStep = 3;
	int freqSave = 1;
	double cfl = 0.95;
	double maxTime = 5;
	double precision = 1e-8;

	myProblem.setCFL(cfl);
	myProblem.setPrecision(precision);
	myProblem.setMaxNbOfTimeStep(MaxNbOfTimeStep);
	myProblem.setTimeMax(maxTime);
	myProblem.setFreqSave(freqSave);
	myProblem.setFileName(fileName);
	bool ok;

	// evolution
	myProblem.initialize();

	ok = myProblem.run();
	if (ok)
		cout << "Simulation "<<fileName<<" is successful !" << endl;
	else
		cout << "Simulation "<<fileName<<"  failed ! " << endl;

	cout << "------------ End of calculation !!! -----------" << endl;
	myProblem.terminate();

	return EXIT_SUCCESS;
}
